File: Matrix-Diagonal-Sum.cpp
1class Solution {
2public:
3    int diagonalSum(vector<vector<int>>& mat) {
4        int sum = 0;
5        int R = mat.size(), C = mat[0].size();
6        for(int i = 0 ; i < R ; i++){
7            for(int j = 0 ; j < C ; j++){
8                if(i==j || (i+j)==R-1) sum+=mat[i][j];
9            }
10        }
11        return sum;
12    }
13};

File: Matrix-Diagonal-Sum.java
1/*
2O(N) solution.
3[r][r] -> primary diagonal
4[R-r-1] -> secondary diagonal 
5
6if matrix is odd length -> middle elem gets added twice
7so remove it 
8
9return ans
10*/
11class Solution {
12    public int diagonalSum(int[][] mat) {
13        int ans = 0;
14        int R = mat.length;
15        for(int r = 0 ; r < R ; r++){
16            ans+=mat[r][r];
17            ans+=mat[R-r-1][r];
18        }
19        if(R%2!=0) ans-=mat[R/2][R/2];
20        return ans;
21    }
22}